### PR TITLE
Track rating

### PR DESCRIPTION
**Pull request by StefanMavrodiev**
Allow users to submit ratings for track versions.

### CODE REVIEW

File: script/migration/migrations/20190818182608_user_rating.php [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+class UserRating extends AbstractMigration
+{
+    public function up()
+    {
+        /**
+         * Create 'rating' table
+         */
+        $this->query("
+            CREATE TABLE rating (
```
Comment (angelk) [inline script/migration/migrations/20190818182608_user_rating.php]: can we rename the table to rating_version?

File: src/Entity/Track/Version.php [hunk 2]
```diff
@@ -142,4 +151,43 @@ public function setNegativeElevation(int $negativeElevation): void
...
+        $total = $this->ratings->count();
+
+        /**
+         * @var $rating Rating
+         */
+        foreach ($this->ratings as $rating) {
+            $sum += $rating->getRating();
+        }
+
+        return round($sum / $total, 1);
```
Comment (angelk) [inline src/Entity/Track/Version.php]: round() should be done in the template.
We may need exact number to order track by rating

File: src/Entity/Track/Rating.php [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+     *     )
+     */
+    private $user;
+
+    /**
+     * @ORM\Column(type="integer")
+     */
+    private $rating;
+
+    public function getId(): ?int
```
Comment (angelk) [inline src/Entity/Track/Rating.php]: guid should be string, not int

File: src/Controller/Track.php [hunk 4]
```diff
@@ -213,6 +217,74 @@ public function newVersion(Request $request, string $id)
...
+            /*
+             * If $rating is null, create new row
+             */
+            if (is_null($rating)) {
+                $rating = new VersionRating();
+                $rating->setUser($user);
+                $rating->setVersion($version);
+            }
+
+            $rating->setRating($request->request->get('rating'));
```
Comment (angelk) [inline src/Controller/Track.php]: `$request->request->get('rating')` should be validated.
At the moment user can alter the html and send ratings like 9, 999, -100

**Comment by angelk** [comment]
@StefanMavrodiev there is 1 minor bug.
The click event for the voting is attached every time a modal is shown.
User click "vote", modal is shown, user vote with 4 stars -> 1 ajax request is send.
User click "vote", modal is shown, user vote with 4 stars -> 2 ajax requests are send.
User click "vote", modal is shown, user vote with 4 stars -> 3 ajax requests are send.
etc...

This doesn't affect the business logic.
We can
1) Deploy it as it is and then fix the issues in another MR
2) Fix the issue in this MR
What do you prefer?

**Comment by StefanMavrodiev** [comment]
You must remove eventListener on modal.bs.hide, e.g. removeEventListener.

For now use jQuery on() function instead of addEventListener.

**Pull request merged**